--- sycl/source/detail/event_impl.hpp	(before formatting)
+++ sycl/source/detail/event_impl.hpp	(after formatting)
@@ -165,9 +165,7 @@
   /// Returns vector of DepDesc dependencies.
   ///
   /// @return a reference to MDeps.
-  std::vector<DepDesc> &getDeps() {
-    return MDeps;
-  }
+  std::vector<DepDesc> &getDeps() { return MDeps; }
 
   /// Returns vector of event dependencies.
   ///
--- sycl/source/detail/scheduler/commands.cpp	(before formatting)
+++ sycl/source/detail/scheduler/commands.cpp	(after formatting)
@@ -333,10 +333,11 @@
 }
 
 Command::Command(CommandType Type, QueueImplPtr Queue)
-    : MQueue(std::move(Queue)), MEvent(std::make_shared<detail::event_impl>(MQueue)),
+    : MQueue(std::move(Queue)),
+      MEvent(std::make_shared<detail::event_impl>(MQueue)),
       MPreparedDepsEvents(MEvent->getPreparedDepsEvents()),
-      MPreparedHostDepsEvents(MEvent->getPreparedHostDepsEvents()),
-      MType(Type), MDeps(MEvent->getDeps()) {
+      MPreparedHostDepsEvents(MEvent->getPreparedHostDepsEvents()), MType(Type),
+      MDeps(MEvent->getDeps()) {
   MSubmittedQueue = MQueue;
   MEvent->setCommand(this);
   MEvent->setContextImpl(MQueue->getContextImplPtr());
